util/sharding_writer/sharding_writer.go

```diff
@@ -0,0 +1,303 @@
+	// If the user sets an extension, prepend the period if they didn't
+	// so we can more simply construct the filename.
+	if len(extension) > 0 && !strings.HasPrefix("extension", ".") {
+		w.extension = fmt.Sprintf(".%s", extension)
+	}
+
+	// Check the opts and set any the user has chosen to set explicitly.
+	for _, o := range opts {
+		o.set(w.opts)
+	}
+
+	if w.opts.shardSize <= 0 {
+		// 2^64 bytes in a file ought to be enough for anybody.
+		w.opts.shardSize = math.MaxInt64
+	}
+
+	var err error
+	w.currentFile, err = os.Create(w.currentFileName())
+	if err != nil {
+		log.Fatalf("unable to create sharded file %q", w.currentFileName())
+	}
+	return w
+}
+
+// currentFileName is a helper to format the current filename.
+func (w *Writer) currentFileName() string {
+	// TODO(rsned): Come to a decision on shard number width and padding.
+	// E.g., %d, %05d, etc.
+	return fmt.Sprintf("%s_%d%s", w.basePath, w.currentShard, w.extension)
+}
+
+// Write writes len(p) bytes from p to the underlying data stream. It returns
+// the number of bytes written from p (0 <= n <= len(p)) and any error encountered
+// that caused the write to stop early.
+func (w *Writer) Write(p []byte) (n int, err error) {
+	if w.errorTriggered {
+		return 0, fmt.Errorf("error occurred during previous writes")
+	}
+
+	// Will this fit without needing to start a new shard?
+	if w.opts.shardSize-w.bytesWritten-int64(len(p)) > 0 {
+		n, err := w.currentFile.Write(p)
+		w.bytesWritten += int64(n)
+		if err != nil || n != len(p) {
+			w.errorTriggered = true
+		}
+		return n, err
+	}
+
+	// TODO(rsned): Implement the sharded writing of bytes.
+	return -1, fmt.Errorf("sharded write not implemented")
+}
+
+// WriteString writes the contents of the given string. It returns the number n of
+// bytes written from from the string (0 <= n <= len(s)) and any error encountered
+// that caused the write to stop early.
+func (w *Writer) WriteString(s string) (n int, err error) {
+	if w.errorTriggered {
+		return 0, fmt.Errorf("error occurred during previous writes")
+	}
+
+	// TODO(rsned): Future enhancement: Allow a +/- 1% leeway so we don't end up
+	// writing out a 100 MB shard and a 1 kB shard when it could all
+	// end up in one file.
+
+	// Check if we can write this without triggering a new shard.
+	if w.opts.shardSize-w.bytesWritten-int64(len(s)) > 0 {
+		n, err := w.currentFile.WriteString(s)
+		w.bytesWritten += int64(n)
+		if err != nil || n != len(s) {
+			w.errorTriggered = true
+		}
+		return n, err
+	}
+
+	var bytesWritten int64
+
+	switch w.opts.dataType {
+	case dataTypeBytes:
+		// We will need more than the current shard to complete the write.
+		for {
+			// Figure out how much of the input we can use before sharding.
+			splitPoint := w.opts.shardSize - w.bytesWritten
+			// If we don't have enough to fill out a shard, adjust the cut point.
+			if splitPoint > int64(len(s)) {
+				splitPoint = int64(len(s))
+			}
+
+			n, err = w.currentFile.WriteString(s[:splitPoint])
```

Comment:
What happens if w.opts.shardSize == w.bytesWritten?  The shard is still open, but it has no space to write more.  Perhaps writing zero bytes will succeed, but it is wasteful.  If should be closed with another opened.  Perhaps move the new shard code up here somewhere.

if len(s) == 0, then stop the iteration.
// Now we know we have something to write.
if the shard is full, then make a new shard.
write to the file
Repeat.